realm/realm-library/src/main/cpp/io_realm_internal_OsObjectStore.cpp

```diff
@@ -140,3 +141,25 @@ JNIEXPORT jboolean JNICALL Java_io_realm_internal_OsObjectStore_nativeDeleteTabl
     CATCH_STD()
     return JNI_FALSE;
 }
+
+JNIEXPORT jboolean JNICALL Java_io_realm_internal_OsObjectStore_nativeCallWithLock(JNIEnv* env, jclass,
+                                                                                   jstring j_realm_path,
+                                                                                   jobject j_runnable)
+{
+    TR_ENTER();
+    try {
+        JStringAccessor path_accessor(env, j_realm_path);
+        std::string realm_path(path_accessor);
+        static JavaClass runnable_class(env, "java/lang/Runnable");
+        static JavaMethod run_method(env, runnable_class, "run", "()V");
+        bool exception_thrown = false;
+        bool result = SharedGroup::call_with_lock(realm_path, [&](std::string path) {
+            REALM_ASSERT_RELEASE_EX(realm_path.compare(path) == 0, realm_path.c_str(), path.c_str());
+            env->CallVoidMethod(j_runnable, run_method);
+            exception_thrown = env->ExceptionCheck();
```

Comment:
This callback is made on the same thread right, so using the `env` should be safe here?

Reply:
yes, it is a synchronized callback. It will return before the outer function returns.